### ISSUE TITLE
Allow the explicit selection of members

### ISSUE THREAD

**Issue by ofek** [pawamoy/mkdocstrings#44]
Often, you'll only want to document a subset of methods & attributes. For example: https://raw.githubusercontent.com/DataDog/integrations-core/hack-a-doc/docs/developer/base/api.md

Is that possible currently? If so, we'll switch to this from `mkautodoc` 😄 

cc @pawamoy

**Comment by pawamoy** [comment]
It is not currently possible, but definitely planned!

I'll add a roadmap to the readme to give a better idea of what I'm planning for mkdocstrings :)

**Comment by ofek** [comment]
Thank you, please let me know if I can help or beta test!

**Comment by pawamoy** [comment]
You'll soon be able to beta-test @ofek! But first I could even use your input!

I can think of several way to select or unselect members:

- exact matches (name)
- regular expressions (name)
- filtering functions (properties)

I'd like to find an elegant way to address all these possibilities, to have something powerful but not too complex, and easy to configure and reason about.

**Comment by ofek** [comment]
I suppose regex is most elegant since they can also represent exact matches

**Comment by pawamoy** [comment]
I personally need to filter out modules in `mkdocstrings` itself, so I was also thinking about something like:

```
::: mkdocstrings.handlers
    selection:
      recurse_on:
        modules: no
```

Maybe a filter could be a dict?

```yaml
filters:
  - categoy: module
    select: no
  - property:
      - private
      - special
    select: no
  - category: method
    name: "^__init__$"
    select: yes
```

**Comment by ofek** [comment]
Seems verbose, though cc @florimondmanca for feedback

**Comment by florimondmanca** [comment]
Oh hi 😄

Ofek's link leads to a Markdown file formatted with [`mkautodoc`](https://github.com/tomchristie/mkautodoc), so…

How about plainly stealing mkautodoc's `:members:` directive? No regular expressions, no include/exclude mechanism, no smarts… Just a "manual mode" switch _first_?

```rst
.. Include all members:
::: BankAccount
    :members:

.. Include only a manual selection of members:
::: BankAccount
    :members: number get_balance compute_vat
```

(`mkdocstrings` doesn't seem to use standard ReStructuredText, but I guess we can adapt this fairly easily.)

 Thoughts? :-)

**Comment by pawamoy** [comment]
> (Note the non-usage of a YAML array, as it would consume a lot of vertical space…)

I'd argue the opposite: long lines are hard to read and generate noise in code diffs :confused: 

I largely prefer

```yaml
::: datadog_checks.base.stubs.aggregator.AggregatorStub
    selection:
      members:
        - assert_metric
        - assert_metric_has_tag
        - assert_metric_has_tag_prefix
        - assert_service_check
        - assert_event
        - assert_all_metrics_covered
        - assert_no_duplicate_metrics
        - assert_no_duplicate_service_checks
        - assert_no_duplicate_all
        - reset
```

over

```yaml
::: datadog_checks.base.stubs.aggregator.AggregatorStub
    selection:
      members: assert_metric assert_metric_has_tag assert_metric_has_tag_prefix assert_service_check assert_event assert_all_metrics_covered assert_no_duplicate_metrics assert_no_duplicate_service_checks assert_no_duplicate_all reset
```

Besides, since we're using YAML, I'd would prefer we stick to it. Otherwise we start doing implicit `split(" ")` and it's an open door to more implicit things which then become hard to maintain :confused: 

Note that in YAML you can still write arrays on one line:

```yaml
::: datadog_checks.base.stubs.aggregator.AggregatorStub
    selection:
      members: [assert_metric, assert_metric_has_tag, assert_metric_has_tag_prefix, assert_service_check, assert_event, assert_all_metrics_covered, assert_no_duplicate_metrics, assert_no_duplicate_service_checks, assert_no_duplicate_all, reset]
```

Anyway, I think it's wise to start simple, with a `members` option and maybe `exclude_members` as well.

**Comment by florimondmanca** [comment]
Sounds good to me @pawamoy!

**Comment by StephenBrown2** [comment]
I'm definitely in favor of a `members` option, whether it's yaml or not doesn't make too much difference, but I especially want to be able to exclude certain members that maybe I didn't create and are inherited from a module's superclass, for example. (`__json_encoder__` from a Pydantic model with custom config, is what I'm dealing with at the moment)

**Comment by chrieke** [comment]
Would the members solution mean that filters would not be available? E.g. would love to filter out private methods (which seem to be included by default in the newest release) without having to manually update the function list often. However not a dealbreaker if not, thank you very much for the great updates!

**Comment by pawamoy** [comment]
Here's what I'm thinking doing:

- in `pytkdocs`, implement powerful filtering like in the last code snippet of this comment https://github.com/pawamoy/mkdocstrings/issues/44#issuecomment-602175240
- in `mkdocstrings`, add the `filters` selection option to reflect `pytkdocs` filtering, and __add a `members` option as well__.
  - If `members` is a list of strings, discard global and local filters, translate the list to exact-match filters internally.
  - If `members` is `yes`, use global and local filters to allow fine-grained selection.
  - If `members` is `no`, select nothing.

Example:

```yaml
# mkdocs.yml

plugins:
  - mkdocstrings:
      # include all members by default
      members: yes
      selection:
        filters:
          # but exclude private members
          - name: "^_[^_]"
            select: no
```

```yaml
# Reference file 

::: pkg.mod.obj1
    selection:
      # explicit selection
      members:
        - member1
        - member2
        - etc


::: pkg.mod.obj2
    selection:
      # no members at all
      members: no


::: pkg.mod.obj3
    selection:
      # global members = yes
      # cancel global filters to include private members as well
      filters: []
```

I think it covers most use-cases?

TODO (if it sounds good):

- [ ] implement this filtering in `pytkdocs`
- [ ] allow setting global `selection` and `rendering` options directly inside `mkdocstrings`' config (not just under `mkdocstrings: handlers: <default_handler>: selection`)
- [ ] add `filters` and `members` option to the Python collector, with respective handling
- [ ] set a default global filter to exclude private members in the Python collector options (to replicate what was previously done :sweat_smile:)?

**Comment by StephenBrown2** [comment]
Sounds good to me! 👍

**Comment by pawamoy** [comment]
It turned out to be much, much more complex than I thought. So I kept things simple and I made progress: `pytkdocs` itself will accept `members` and `filters` options. `members` can be `True`, `False`, or a list of names to select. If it's a list of names, we don't apply filters on the top object's members. We still apply filters recursively, on all selected members and their children. Filters are regular expressions, and can (re)select or unselect (when starting with `!`) objects, so their order matters.

It's not as powerful as I wanted it to be, but who cares if the use-cases are covered :slightly_smiling_face: 

Also changes were minimal :+1:

**Comment by ofek** [comment]
Yay! Can we try it yet?

**Comment by pawamoy** [comment]
You can now :slightly_smiling_face:  I just released version 0.10.0. And I forgot to implement the second point of the checklist above :confused: Will do in the next one!

**Comment by ofek** [comment]
Nice, thanks!!! Found a few things:

https://github.com/pawamoy/mkdocstrings/issues/64
https://github.com/pawamoy/mkdocstrings/issues/65
